### PR TITLE
move the `message.Apply` to the last call / change parameter name of `RequestOptions` to `options` in non-branded case

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes https://github.com/Azure/autorest.csharp/issues/4484
Fixes https://github.com/Azure/autorest.csharp/issues/4482

# Description
<i>Add your description here!</i>

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs:32]: parameter `methodAccessibility` is removed here because in all the places we are invoking this method, we always pass in the value "internal".
Therefore for simplicity, I removed it and hard-coded it in the method.

We should schedule a refactor to this method to use the expressions to build the method instead of using strings, and when we do that, we could add the parameter back via `MethodSignatureModifiers`

File: src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs:31]: I change the parameter `fields` to non-nullable because in all the use case of this method, it is never null to reduce complexity.

File: src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs:257]: make `fields` non-nullable because it is never null

File: src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs:301]: make `fields` non-nullable because it is never null

File: src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs:338]: make `fields` non-nullable because it is never null

File: src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs:338]: removed because it is never used.

**Comment by ArcturusZhang** [comment]
@m-nash @annelo-msft  could you please have a review?

**Review by chunyu3 — APPROVED** [review]

**Review by m-nash — APPROVED** [review]
LGTM @annelo-msft do the resulting client changes look correct?

File: src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs]: Can you link a tracking issue for this?

File: test/UnbrandedProjects/NoTest-TypeSpec/src/Generated/NoTestTypeSpecClient.cs
Comment (annelo-msft) [inline test/UnbrandedProjects/NoTest-TypeSpec/src/Generated/NoTestTypeSpecClient.cs:166]: Why do we need the null check?  I think it is better to throw here to indicate an error.  If we instead don't apply the options to the message, we could get strange behavior from the pipeline without having identified the root of the problem -- it will make it harder to find bugs in either client or user code.

File: test/CadlRanchProjectsNonAzure/authentication/http/custom/src/Generated/CustomClient.cs
Comment (annelo-msft) [inline test/CadlRanchProjectsNonAzure/authentication/http/custom/src/Generated/CustomClient.cs:69]: In protocol methods where options can be null or default, we'll need to add this line:

```
options ??= new RequestOptions();
```

You can see https://github.com/annelo-msft/openai-in-typespec/pull/3/files#diff-e44e45d43a3dd8ac6d3c5c5fc6cacb2ce2b5f138500a60cc98237a4500828f89R86-R93 for details.

File: test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/AudioTranscriptions.cs
Comment (annelo-msft) [inline test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/AudioTranscriptions.cs:51]: Not sure if you plan to handle this in this or a later PR, but:
1. Convenience methods in unbranded clients should not take CancellationTokens
2. Once 1. is addressed, you can remove the `FromCancellationToken` helper method from unbranded clients

File: test/UnbrandedProjects/NoTest-TypeSpec/src/Generated/NoTestTypeSpecClient.cs
Comment (m-nash) [reply test/UnbrandedProjects/NoTest-TypeSpec/src/Generated/NoTestTypeSpecClient.cs:166]: It's because parameter from the public API is optional so this can be null here.

File: test/CadlRanchProjectsNonAzure/authentication/http/custom/src/Generated/CustomClient.cs
Comment (m-nash) [reply test/CadlRanchProjectsNonAzure/authentication/http/custom/src/Generated/CustomClient.cs:69]: I don't think we should be instantiating an object if we don't need that object.  It's just extra memory for now reason.

File: test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/AudioTranscriptions.cs
Comment (m-nash) [reply test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/AudioTranscriptions.cs:51]: Yes remove cancellation token is tracked here https://github.com/Azure/autorest.csharp/issues/4243

File: test/CadlRanchProjectsNonAzure/authentication/http/custom/src/Generated/CustomClient.cs
Comment (annelo-msft) [reply test/CadlRanchProjectsNonAzure/authentication/http/custom/src/Generated/CustomClient.cs:69]: @m-nash and I discussed this offline.  I see the concern and I think it's probably OK to do it the way it's proposed in this PR.  I want to double-check with @KrzysztofCwalina that this approach doesn't cause problems for the RequestOptions freezing story... I can't think offhand of how it would, but don't know if there is a case I haven't considered here.

We'll talk to @KrzysztofCwalina about it in the Azure.Core meeting on Thursday, and can loop back then.  Since I think it's reasonably low-risk, let's not block this PR on the question and we can update the generated code if we learn we need to.

File: test/UnbrandedProjects/NoTest-TypeSpec/src/Generated/NoTestTypeSpecClient.cs
Comment (annelo-msft) [reply test/UnbrandedProjects/NoTest-TypeSpec/src/Generated/NoTestTypeSpecClient.cs:166]: Same response as https://github.com/Azure/autorest.csharp/pull/4502#discussion_r1548462971

File: test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/AudioTranscriptions.cs
Comment (annelo-msft) [reply test/UnbrandedProjects/Platform-OpenAI-TypeSpec/src/Generated/AudioTranscriptions.cs:51]: Sounds good.

File: test/UnbrandedProjects/NoTest-TypeSpec/src/Generated/NoTestTypeSpecClient.cs
Comment (ArcturusZhang) [reply test/UnbrandedProjects/NoTest-TypeSpec/src/Generated/NoTestTypeSpecClient.cs:166]: @annelo-msft please create an issue for this if we plan to change this.
I will resolve it for now - but if we decide to change this, we could do that as follow ups

File: test/CadlRanchProjectsNonAzure/authentication/http/custom/src/Generated/CustomClient.cs
Comment (ArcturusZhang) [reply test/CadlRanchProjectsNonAzure/authentication/http/custom/src/Generated/CustomClient.cs:69]: please create issues if follow ups are needed. I would like to resolve these to finish this PR to unblock others, thank you

File: src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs
Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs]: I believe it is tracking here by Alexander: https://github.com/Azure/autorest.csharp/issues/3768

**Pull request merged**
